File: Analysis/Pipeline/TrainingScriptsT1Only/batchOctantGenerator.R
batchGenerator <- function( batchSize = 32L,
                            patchSize = c( 64, 64, 64 ),
                            template = NULL,
                            t1s = NULL,
                            flairs = NULL,
                            brainMasks = NULL,
                            segmentationImages = NULL,
			                      segmentationLabels = NULL,
                            doRandomContralateralFlips = TRUE,
                            doDataAugmentation = TRUE
 )
{                                     

  if( is.null( template ) )
    {
    stop( "No reference template specified." )
    }
  if( is.null( t1s ) || is.null( flairs ) )
    {
    stop( "Input images must be specified." )
    }
  if( is.null( segmentationImages ) )
    {
    stop( "Input masks must be specified." )
    }
  if( is.null( segmentationLabels ) )
    {
    stop( "segmentationLabels must be specified." )
    }

  strideLength <- dim( template ) - patchSize

  currentPassCount <- 0L

  function()
    {
    # Shuffle the data after each complete pass

    if( ( currentPassCount + batchSize ) >= length( t1s ) )
      {
      # shuffle the source data
      sampleIndices <- sample( length( t1s ) )
      t1s <- t1s[sampleIndices]
      flairs <- flairs[sampleIndices]
      brainMasks <- brainMasks[sampleIndices]
      segmentationImages <- segmentationImages[sampleIndices]

      currentPassCount <- 0L
      }

    batchIndices <- currentPassCount + 1L:batchSize

    batchT1s <- t1s[batchIndices]
    batchFlairs <- flairs[batchIndices]
    batchBrainMasks <- brainMasks[batchIndices]
    batchWmhs <- segmentationImages[batchIndices]

    X <- array( data = 0, dim = c( batchSize, patchSize, 1 ) )    
    Y <- array( data = 0L, dim = c( batchSize, patchSize ) )

    currentPassCount <<- currentPassCount + batchSize

    # pb <- txtProgressBar( min = 0, max = batchSize, style = 3 )
    
    count <- 1
    while( count <= batchSize )
      {
      i <- sample.int( batchSize, 1 ) 	      
  
      # setTxtProgressBar( pb, i )

      t1 <- NULL 
      flair <- NULL
      wmh <- NULL

      mask <- antsImageRead( batchBrainMasks[i] )
      batchT1 <- antsImageRead( batchT1s[i] ) * mask
      batchFlair <- antsImageRead( batchFlairs[i] ) * mask
      batchWmh <- antsImageRead( batchWmhs[i] )

      # antsImageWrite( batchT1, "t1.nii.gz" );
      # antsImageWrite( batchFlair, "flair.nii.gz" );
      # antsImageWrite( batchWmh, "wmh.nii.gz" );
      # stop( "HERE" )

      if( doRandomContralateralFlips && sample( c( TRUE, FALSE ) ) )
        {
        t1A <- as.array( batchT1 )
	t1 <- as.antsImage( t1A[dim( t1A )[1]:1,,], reference = batchT1 )
        flairA <- as.array( batchFlair )
        flair <- as.antsImage( flairA[dim( flairA )[1]:1,,], reference = batchFlair )
        wmhA <- as.array( batchWmh )
        wmh <- as.antsImage( wmhA[dim( wmhA )[1]:1,,], reference = batchWmh )
        } else {
        t1 <- batchT1  
	      flair <- batchFlair
        wmh <- batchWmh
        }
      
      warpedT1 <- t1
      warpedFlair <- flair
      warpedWmh <- wmh
      
      wmhPatches <- NULL
      t1Patches <- NULL
      flairPatches <- NULL

      if( doDataAugmentation == TRUE )
        {
        dataAugmentation <- 
          randomlyTransformImageData( template, 
          list( list( warpedT1, warpedFlair ) ),
          list( warpedWmh ),
          numberOfSimulations = 1, 
          transformType = 'affineAndDeformation', 
          sdAffine = 0.01,
          deformationTransformType = "bspline",
          numberOfRandomPoints = 1000,
          sdNoise = 2.0,
          numberOfFittingLevels = 4,
          meshSize = 1,
          sdSmoothing = 4.0,
          inputImageInterpolator = 'linear',
          segmentationImageInterpolator = 'nearestNeighbor' )

        simulatedT1 <- dataAugmentation$simulatedImages[[1]][[1]]
        simulatedT1 <- ( simulatedT1 - mean( simulatedT1 ) ) / sd( simulatedT1 )
        simulatedFlair <- dataAugmentation$simulatedImages[[1]][[2]]
        simulatedFlair <- ( simulatedFlair - mean( simulatedFlair ) ) / sd( simulatedFlair )

        simulatedWmh <- dataAugmentation$simulatedSegmentationImages[[1]]

        # antsImageWrite( simulatedImage, paste0( "./TempData/simBrainImage", i, ".nii.gz" ) )
        # antsImageWrite( simulatedSegmentationImage, paste0( "./TempData/simSegmentationImage", i, ".nii.gz" ) )

        t1Patches <- extractImagePatches( simulatedT1, patchSize, maxNumberOfPatches = "all",
					  strideLength = strideLength, randomSeed = NULL,
					  returnAsArray = TRUE )
        flairPatches <- extractImagePatches( simulatedFlair, patchSize, maxNumberOfPatches = "all",
                                             strideLength = strideLength, randomSeed = NULL,
                                             returnAsArray = TRUE )
        wmhPatches <- extractImagePatches( simulatedWmh, patchSize, maxNumberOfPatches = "all",
                                           strideLength = strideLength, randomSeed = NULL,
                                           returnAsArray = TRUE )	

        } else {
        warpedT1 <- ( warpedT1 - mean( warpedT1 ) ) / sd( warpedT1 )
        warpedFlair <- ( warpedFlair - mean( warpedFlair ) ) / sd( warpedFlair )

        # antsImageWrite( warpedImage, paste0( "./TempData/wBrainImage", i, ".nii.gz" ) )
        # antsImageWrite( warpedSegmentationImage, paste0( "./TempData/wSegmentationImage", i, ".nii.gz" ) )

        t1Patches <- extractImagePatches( warpedT1, patchSize, maxNumberOfPatches = "all",
                                          strideLength = strideLength, randomSeed = NULL,
                                          returnAsArray = TRUE )
        flairPatches <- extractImagePatches( warpedFlair, patchSize, maxNumberOfPatches = "all",
                                             strideLength = strideLength, randomSeed = NULL,
                                             returnAsArray = TRUE )
        wmhPatches <- extractImagePatches( warpedWmh, patchSize, maxNumberOfPatches = "all",
                                           strideLength = strideLength, randomSeed = NULL,
                                           returnAsArray = TRUE )
      	}

      whichOctant <- sample.int( 8, 1 )
      # cat( i, ", ", whichOctant, ":")  
      if( sum( wmhPatches[whichOctant,,,] ) >= 1000 )
        {
        # cat( "  Yes!! \n")  
        X[count,,,,1] <- t1Patches[whichOctant,,,]
        # X[count,,,,2] <- flairPatches[whichOctant,,,]
        Y[count,,,] <- wmhPatches[whichOctant,,,]

        # antsImageWrite( as.antsImage( drop( t1Patches[whichOctant,,,] ) ), "t1.nii.gz" );
        # antsImageWrite( as.antsImage( drop( flairPatches[whichOctant,,,] ) ), "flair.nii.gz" );
        # antsImageWrite( as.antsImage( drop( wmhPatches[whichOctant,,,] ) ), "wmh.nii.gz" );
        # stop( "HERE" )

        count <- count + 1  
        } # else { 
        # cat( "  No!! \n")  
        # }
      }
    # stop( "Done testing.")  
    # cat( "\n" )

    encodedY <- encodeUnet( Y, segmentationLabels )

    return( list( X, encodedY ) )
    }
}

File: Analysis/Pipeline/TrainingScripts/batch_octant_generator.py
import numpy as np
import random
import ants
import antspynet

def batch_generator(batch_size=32,
                    patch_size=(64, 64, 64),
                    template=None,
                    t1s=None,
                    flairs=None,
                    brain_masks=None,
                    segmentation_images=None,
                    segmentation_labels=None,
                    do_random_contralateral_flips=True,
                    do_data_augmentation=True):

   if template is None:
       raise ValueError("No reference template specified.")

   if t1s is None or flairs is None:
       raise ValueError("Input images must be specified.")

   if segmentation_images is None:
       raise ValueError("Input masks must be specified.")

   if segmentation_labels is None:
       raise ValueError("segmentation labels must be specified.")

   stride_length = tuple(np.subtract(template.shape, patch_size))

   while True:

       X = np.zeros((batch_size, *patch_size, 2))
       Y = np.zeros((batch_size, *patch_size))

       batch_count = 0

       while batch_count < batch_size:
           i = random.sample(range(len(t1s)), 1)[0]

           mask = ants.image_read(brain_masks[i])
           batch_wmh = ants.image_read(segmentation_images[i])

           batch_t1 = ants.image_read(t1s[i]) * mask
           batch_flair = ants.image_read(flairs[i]) * mask

           if do_random_contralateral_flips and random.sample((True, False), 1)[0]:
               t1A = t1.numpy()
               t1 = ants.from_numpy(t1A[t1A.shape[0]::0,:,:])
               flairA = flair.numpy()
               flair = ants.from_numpy(flairA[flairA.shape[0]::0,:,:])
               wmhA = batch_wmh.numpy()
               wmh = ants.from_numpy(wmhA[wmhA.shape[0]::0,:,:])
           else:
               t1 = batch_t1
               flair = batch_flair
               wmh = batch_wmh

           warped_t1 = t1
           warped_flair = flair
           warped_wmh = wmh

           wmh_patches = None
           t1_patches = None
           flair_patches = None

           if do_data_augmentation == True:
               data_augmentation = antspynet.randomly_transform_image_data(template,
                   [[warped_t1, warped_flair]],
                   [warped_wmh],
                   number_of_simulations=1,
                   transform_type='affineAndDeformation',
                   sd_affine=0.01,
                   deformation_transform_type="bspline",
                   number_of_random_points=1000,
                   sd_noise=2.0,
                   number_of_fitting_levels=4,
                   mesh_size=1,
                   sd_smoothing=4.0,
                   input_image_interpolator='linear',
                   segmentation_image_interpolator='nearestNeighbor')

               simulated_t1 = data_augmentation['simulated_images'][0][0]
               simulated_t1 = (simulated_t1 - simulated_t1.mean()) / simulated_t1.std()
               simulated_flair = data_augmentation['simulated_images'][0][1]
               simulated_flair = (simulated_flair - simulated_flair.mean()) / simulated_flair.std()

               simulated_wmh = data_augmentation['segmentation_images'][0]

               t1_patches = antspynet.extract_image_patches(simulated_t1, patch_size, max_number_of_patches='all',
                   stride_length=stride_length, random_seed=None, return_as_array=True)
               flair_patches = antspynet.extract_image_patches(simulated_flair, patch_size, max_number_of_patches='all',
                   stride_length=stride_length, random_seed=None, return_as_array=True)
               wmh_patches = antspynet.extract_image_patches(simulated_wmh, patch_size, max_number_of_patches='all',
                   stride_length=stride_length, random_seed=None, return_as_array=True)

           else:

               warped_t1 = (warped_t1 - warped_t1.mean()) / warped_t1.std()
               warped_flair = (warped_flair - warped_flair.mean()) / warped_flair.std()

               t1_patches = antspynet.extract_image_patches(warped_t1, patch_size, max_number_of_patches='all',
                   stride_length=stride_length, random_seed=None, return_as_array=True)
               flair_patches = antspynet.extract_image_patches(warped_flair, patch_size, max_number_of_patches='all',
                   stride_length=stride_length, random_seed=None, return_as_array=True)
               wmh_patches = antspynet.extract_image_patches(warped_wmh, patch_size, max_number_of_patches='all',
                   stride_length=stride_length, random_seed=None, return_as_array=True)

           which_octant = random.sample(range(8), 1)[0]

           if wmh_patches[which_octant,:,:,:].sum() >= 1000:
               X[batch_count,:,:,:,0] = t1_patches[which_octant,:,:,:]
               X[batch_count,:,:,:,1] = flair_patches[which_octant,:,:,:]
               Y[batch_count,:,:,:] = wmh_patches[which_octant,:,:,:]

               batch_count = batch_count + 1

       encoded_Y = antspynet.encode_unet(Y, segmentation_labels)

       yield X, encoded_Y, [None]









